internal/singleflight/singleflight/singleflight.go

```diff
@@ -0,0 +1,68 @@
+//
+// Copyright (C) 2019-2020 Vdaas.org Vald team ( kpango, rinx, kmrmt )
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+// Package singleflight represents zero time caching.
+package singleflight
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+)
+
+type call struct {
+	wg   sync.WaitGroup
+	val  interface{}
+	err  error
+	dups uint64
+}
+
+// Group represents interface for zero time cache.
+type Group interface {
+	Do(ctx context.Context, key string, fn func() (interface{}, error)) (v interface{}, shared bool, err error)
+}
+
+type group struct {
+	m sync.Map
+}
+
+// New returns Group implementation.
+func New() Group {
+	return new(group)
+}
+
+// Do returns a set of the cache of the first return value from function
+// as interface{}, shared flg as bool, and err as error
+// when the function is called multiple times in an instant.
+func (g *group) Do(ctx context.Context, key string, fn func() (interface{}, error)) (v interface{}, shared bool, err error) {
```

Comment:
I think the `v interface{}` and `err error` are from the `fn func()`, I think we should group them.
Also the standard `singleflight` definition is `func (g *Group) Do(key string, fn func() (interface{}, error)) (v interface{}, err error, shared bool)`. 
https://pkg.go.dev/golang.org/x/sync/singleflight?tab=doc#Group.Do
```suggestion
func (g *group) Do(ctx context.Context, key string, fn func() (interface{}, error)) (v interface{}, err error, shared bool) {
```

Reply:
I can understand your opinion.
I changed to this style before.
This is because go test cannot generate test code. I had a previous discussion and asked them to fix it in the current format.

Reply:
![badge](https://img.shields.io/badge/review-imo-orange.svg?style=flat-square)
I think the first priority is a better implementation, not a compatibility with a generator tool.

Reply:
@hlts2 @kevindiu ping

Reply:
Sorry. 
But I think go standard returns an error at the end.
so I think it's okay to keep it as it is. What do you think?

Reply:
@hlts2 @kevindiu As Kevin thinks, i also think that it is not a good manner to implement a different interface for a function with same name and role.
Also, it is important to indicate the `err` is from the passed function `fn` clearly.

Reply:
It's okay to keep it as it is now, however i recommend to use the same interface as the x/sync/singleflight package.

Reply:
@rinx @kevindiu 
Thank you for your suggestion.
sure, I will fix from `interface{}, bool, error` to `interface{}, error, bool`.:bow:

Reply:
@rinx @kevindiu 
The fix is ​​now complete.